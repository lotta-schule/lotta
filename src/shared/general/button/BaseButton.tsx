import * as React from 'react';
import clsx from 'clsx';

export type BaseButtonProps = {
    /**
     * Wether the button is currently 'active' or 'selected'
     */
    selected?: boolean;

    /**
     * Chose from different styles
     */
    variant?: 'default' | 'fill' | 'borderless' | 'error';

    /**
     * Wether the button should have a width of 100%
     */
    fullWidth?: boolean;

    /**
     * CSS Classname
     */
    className?: string;

    /**
     * CSS Styles
     */
    style?: React.CSSProperties;

    /**
     * Content to show on the button
     */
    children?: any;
} & React.HTMLProps<HTMLButtonElement>;

/**
 * Primary UI shared for userAvatar interaction
 */
export const BaseButton = React.forwardRef<any, BaseButtonProps>(
    (
        {
            children,
            style,
            className,
            variant = 'default',
            fullWidth = false,
            selected = false,
            ...props
        },
        ref
    ) => {
        const ComponentClass = props.href ? 'a' : ('button' as any);
        return React.createElement(
            ComponentClass,
            {
                ref,
                type: props.type ?? 'button',
                role: 'button',
                style,
                ['aria-selected']: selected,
                className: clsx(
                    'lotta-base-button',
                    `variant__${variant}`,
                    { selected, 'full-width': fullWidth },
                    className
                ),
                ...props,
            },
            children
        );
    }
);
BaseButton.displayName = 'BaseButton';
