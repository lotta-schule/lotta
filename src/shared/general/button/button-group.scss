.lotta-button-group {
    display: inline-flex;
    flex-direction: row;
    justify-content: stretch;
    button {
        color: rgba(var(--lotta-primary-color), 1);
        background-color: #fff;
        display: inline-block;
        &:not(:only-child) {
            &:not(:first-of-type):not(:last-of-type):not(.is-first-button-group-button):not(.is-last-button-group-button) {
                border-radius: 0;
                border-right: 0;
            }
            &:first-of-type:not(.is-last-button-group-button),
            &.is-first-button-group-button {
                border-top-right-radius: 0;
                border-bottom-right-radius: 0;
                border-right: 0;
            }
            &:last-of-type:not(.is-first-button-group-button),
            &.is-last-button-group-button {
                border-top-left-radius: 0;
                border-bottom-left-radius: 0;
            }
        }
        &.is-first-button-group-button {
            border-top-right-radius: 0;
            border-bottom-right-radius: 0;
            border-right: 0;
        }
        &.is-last-button-group-button {
            border-top-left-radius: 0;
            border-bottom-left-radius: 0;
        }
        &.full-width {
            flex-grow: 1;
        }
        .lotta-button_icon svg {
            font-size: 1.3rem;
        }
    }
    &.full-width {
        width: 100%;
    }
}
