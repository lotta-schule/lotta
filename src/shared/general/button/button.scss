.lotta-button {
    color: rgba(var(--lotta-primary-color), 1);
    font-weight: 200;
    padding: 0.4em 0.8em;
    align-items: center;
    &:not(.small) {
        min-height: 2.8em;
    }
}

.lotta-button_icon {
    display: flex;
}

.lotta-button_text {
    margin: auto;
    display: contents;
}

aside .lotta-button_text {
    margin: 0;
}

aside .lotta-navigation-button {
    border-radius: 50%;
}

aside nav.lotta-navbar ul {
    height: auto;
}

.lotta-button.only-icon:not(.lotta-button-group .lotta-button) {
    border-radius: 50%;
    border-color: transparent;
    padding: 0.4em;
}

.lotta-button.only-icon
    :not(.lotta-button-group .lotta-button)
    :not(.lotta-button.variant__fill)
    :hover {
    background-color: rgba(var(--lotta-primary-color), 0.2);
}
