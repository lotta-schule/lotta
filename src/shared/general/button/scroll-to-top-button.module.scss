.root {
  position: fixed !important;
  bottom: .5em;
  right: .5em;
}
