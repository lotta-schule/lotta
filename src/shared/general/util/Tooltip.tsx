import * as React from 'react';
import { Tooltip as ReachTooltip } from '@reach/tooltip';
import clsx from 'clsx';

import styles from './Tooltip.module.scss';

export interface TooltipProps extends React.HTMLAttributes<HTMLDivElement> {
    label: React.ReactNode;
}

export const Tooltip = React.forwardRef(
    (
        { children, label, className, ...props }: TooltipProps,
        ref: React.ForwardedRef<HTMLDivElement>
    ) => {
        return (
            <ReachTooltip
                ref={ref}
                label={label}
                className={clsx(styles.root, className)}
                {...props}
            >
                {children}
            </ReachTooltip>
        );
    }
);
Tooltip.displayName = 'Tooltip';
