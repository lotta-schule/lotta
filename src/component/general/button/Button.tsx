import * as React from 'react';
import { BaseButton, BaseButtonProps } from './BaseButton';
import clsx from 'clsx';

export type ButtonProps = {
    /**
     * Button contents
     */
    label?: string;

    /**
     * Icon to show on the button side
     */
    icon?: React.ReactElement;

    /**
     * Disable setting a minimal height
     */
    small?: boolean;

    /**
     * Force style for button with only icon and without label,
     * even if child is found
     */
    onlyIcon?: boolean;
} & BaseButtonProps;

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
    ({ icon, label, onlyIcon, small, children, ...props }, ref) => {
        return (
            <BaseButton
                {...props}
                ref={ref}
                className={clsx('lotta-button', props.className, {
                    'only-icon': onlyIcon || (icon && !(label || children)),
                    small: small,
                })}
            >
                {icon && <span className={'lotta-button_icon'}>{icon}</span>}
                {(label ?? children) && (
                    <span className={'lotta-button_text'}>
                        {label ?? children}
                    </span>
                )}
            </BaseButton>
        );
    }
);
Button.displayName = 'Button';
