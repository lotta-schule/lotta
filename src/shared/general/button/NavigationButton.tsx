import * as React from 'react';
import { Button, ButtonProps } from './Button';
import clsx from 'clsx';

export type NavigationButtonProps = ButtonProps;

export const NavigationButton = React.forwardRef<
    HTMLButtonElement,
    NavigationButtonProps
>((props, ref) => {
    return (
        <Button
            {...props}
            ref={ref}
            className={clsx('lotta-navigation-button', props.className)}
        />
    );
});
NavigationButton.displayName = 'NavigationButton';
