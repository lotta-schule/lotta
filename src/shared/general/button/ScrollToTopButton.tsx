import * as React from 'react';
import { Button } from './Button';
import { ArrowUpwardRounded } from '@material-ui/icons';
import { useSpring, animated } from 'react-spring';
import { useScrollEvent } from 'util/useScrollEvent';
import { useWindowSize } from 'util/useWindowSize';

import styles from './scroll-to-top-button.module.scss';

const AnimatedButton = animated(Button);

export const ScrollToTopButton = React.memo(() => {
    const [isShown, setIsShown] = React.useState(false);

    const { innerHeight } = useWindowSize();
    const onScroll = React.useCallback(() => {
        setIsShown(window.scrollY > 2 * innerHeight);
    }, [innerHeight]);
    useScrollEvent(onScroll, 1000, [onScroll]);

    const props = useSpring({
        opacity: isShown ? 1 : 0,
        tension: 2000,
        mass: 0.5,
        friction: 15,
    });
    return (
        <AnimatedButton
            style={props}
            className={styles.root}
            title={'Zum Anfang der Seite scrollen'}
            icon={<ArrowUpwardRounded />}
            onClick={() => {
                window.scroll({
                    top: 0,
                    behavior: 'smooth',
                });
            }}
        />
    );
});
ScrollToTopButton.displayName = 'ScrollToTopButton';
