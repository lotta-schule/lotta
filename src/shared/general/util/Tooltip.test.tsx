import { render } from '@testing-library/react';
import { Button } from '../button/Button';
import { Tooltip } from './Tooltip';

describe('util/Tooltip', () => {
    it('should render', () => {
        const screen = render(
            <Tooltip label="Test">
                <Button>Test</Button>
            </Tooltip>
        );
        expect(screen.container).toMatchSnapshot();
    });
});
