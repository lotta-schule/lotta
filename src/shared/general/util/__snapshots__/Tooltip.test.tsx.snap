// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util/Tooltip should render 1`] = `
<div>
  <button
    aria-selected="false"
    class="lotta-base-button variant__default lotta-button"
    data-reach-tooltip-trigger=""
    data-state="tooltip-hidden"
    role="button"
    type="button"
  >
    <span
      class="lotta-button_text"
    >
      Test
    </span>
  </button>
</div>
`;
